File: 0538.把二叉搜索树转换为累加树/0538-把二叉搜索树转换为累加树.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var convertBST = function(root) {
    let accumulator = 0;
    var dfs = (root) => {
        if(root == null) {
            return
        }
        dfs(root.right);
        accumulator += root.val;
        root.val = accumulator;
        dfs(root.left);
    };
    dfs(root);
    return root;
};

File: 0538.把二叉搜索树转换为累加树/0538-把二叉搜索树转换为累加树.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {TreeNode}
def convert_bst(root)
    @accumulator = 0
    dfs(root)
end

def dfs(root)
    return nil if root.nil?
    dfs(root.right)
    @accumulator += root.val
    root.val = @accumulator
    dfs(root.left)
    root
end